### PR TITLE
Enable in memory project files

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes

- Enable using options from project files that are opened but not saved
- Ignore project files whose includes/excludes do not match the current file

### Testing
- Added a unit test for each of the above changes

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Options/InMemoryDirectoryInfoFromDotNet8.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.FileSystemGlobbing.Abstractions;
+
+namespace DafnyCore.Options;
+
+/// <summary>
+/// Contains a bugfix that's available in DotNet8
```
Comment (robin-aws) [inline Source/DafnyCore/Options/InMemoryDirectoryInfoFromDotNet8.cs:11]: Could we just require .NET 8 now instead of copying this code?

**Review by robin-aws — COMMENTED** [review]
Only skimmed for now, had a quick question

File: Source/DafnyCore/Options/InMemoryDirectoryInfoFromDotNet8.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/Options/InMemoryDirectoryInfoFromDotNet8.cs:11]: It's in preview

File: Source/DafnyLanguageServer/Workspace/CombinedDirectoryInfo.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.FileSystemGlobbing.Abstractions;
+
+namespace Microsoft.Dafny.LanguageServer.Workspace;
+
+class CombinedDirectoryInfo : DirectoryInfoBase {
```
Comment (alex-chew) [inline Source/DafnyLanguageServer/Workspace/CombinedDirectoryInfo.cs:9]: Same comment as in #3902 - please document what this class does.

**Review by alex-chew — DISMISSED** [review]
LGTM

File: Source/DafnyLanguageServer.Test/Util/ClientBasedLanguageServerTest.cs [hunk 3]
```diff
@@ -159,7 +160,8 @@ public class ClientBasedLanguageServerTest : DafnyLanguageServerTestBase, IAsync
 
       }
     }
-    var verificationDocumentItem = CreateTestDocument("class X {does not parse", $"AssertNoDiagnosticsAreComing{fileIndex++}.dfy");
+    var verificationDocumentItem = CreateTestDocument("class X {does not parse",
+      Path.Combine(Directory.GetCurrentDirectory(), $"AssertNoDiagnosticsAreComing{fileIndex++}.dfy"));
```
Comment (alex-chew) [inline Source/DafnyLanguageServer.Test/Util/ClientBasedLanguageServerTest.cs:164]: nit: it's a bit unfortunate to have tests that depend on the working directory

**Review by stefan-aws — APPROVED** [review]
Pre-approved modulo small fixes.

**Pull request merged**